public/app/plugins/panel/logs/LogsPanel.tsx

```diff
@@ -294,10 +368,59 @@ async function copyDashboardUrl(row: LogRowModel, timeRange: TimeRange) {
 
   // Add panel state containing the rowId, and absolute time range from the current query, but leave everything else the same, if the user is in edit mode when grabbing the link, that's what will be linked to, etc.
   currentURL.searchParams.set('panelState', JSON.stringify(panelState));
-  currentURL.searchParams.set('from', toUtc(timeRange.from).valueOf().toString(10));
-  currentURL.searchParams.set('to', toUtc(timeRange.to).valueOf().toString(10));
+  const range = getPermalinkRange(row, rows, {
+    from: toUtc(timeRange.from).valueOf(),
+    to: toUtc(timeRange.to).valueOf(),
+  });
+  currentURL.searchParams.set('from', range.from.toString(10));
+  currentURL.searchParams.set('to', range.to.toString(10));
 
   await createAndCopyShortLink(currentURL.toString());
 
   return Promise.resolve();
 }
+
+async function requestMoreLogs(panelData: PanelData, timeRange: AbsoluteTimeRange, timeZone: TimeZone) {
+  if (!panelData.request) {
+    return [];
+  }
+
+  const range: TimeRange = convertRawToRange({
+    from: dateTimeForTimeZone(timeZone, timeRange.from),
+    to: dateTimeForTimeZone(timeZone, timeRange.to),
+  });
+
+  const targetGroups = groupBy(panelData.request.targets, 'datasource.uid');
+  const dataRequests = [];
+
+  for (const uid in targetGroups) {
+    const dataSource = await getDataSourceSrv().get(uid);
+    if (!dataSource) {
+      continue;
+    }
```

Comment:
It's possible that I need to sync this with recent changes to this component.

Reply:
I think this comment is still actual. Any reason not to use `dataSourcesMap`?

Reply:
I forgot about this comment, thanks for the reminder! I updated the code, although it would have been slightly more convenient that the hook returns a map of uid => datasource instead of refId => datasource, which returns the same instance for multiple targets, so instead of map[uid] you have to map[targets[0].refId]